bin/pg_repack.c

```diff
@@ -2073,14 +2113,47 @@ repack_all_indexes(char *errbuf, size_t errsize)
 
 		cell = r_index.head;
 	}
-	else if (table_list.head)
+	else if (table_list.head || parent_table_list.head)
 	{
 		appendStringInfoString(&sql,
 			"SELECT i.relname, idx.indexrelid, idx.indisvalid, idx.indrelid, $1::text, n.nspname"
 			" FROM pg_index idx JOIN pg_class i ON i.oid = idx.indexrelid"
 			" JOIN pg_namespace n ON n.oid = i.relnamespace"
 			" WHERE idx.indrelid = $1::regclass ORDER BY indisvalid DESC, i.relname, n.nspname");
 
+		for (cell = parent_table_list.head; cell; cell = cell->next)
+		{
+			int nchildren, i;
+
+			params[0] = cell->val;
+
+			/* find children of this parent table */
+			res = execute_elevel("SELECT quote_ident(n.nspname) || '.' || quote_ident(c.relname)"
+								 " FROM pg_class c JOIN pg_namespace n on n.oid = c.relnamespace"
+								 " WHERE c.oid = ANY (repack.get_table_and_inheritors($1::regclass))"
+								 " ORDER BY n.nspname, c.relname", 1, params, DEBUG2);
+
+			if (PQresultStatus(res) != PGRES_TUPLES_OK)
+			{
+				elog(WARNING, "%s", PQerrorMessage(connection));
+				continue;
+			}
+
+			nchildren = PQntuples(res);
+
+			if (nchildren == 0)
```

Comment:
@MasahikoSawada, actually, I've inspected `regclassin` and it seems that casts to `regclass` type don't take any locks on relations.

As far as I understand, pg_repack doesn't take locks in `repack_one_database()`. Here's a test:

Step 1:
```bash
# terminal A
gdb --args ./bin/pg_repack -t test db
...
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./bin/pg_repack...done.

(gdb) b repack_one_table 
Breakpoint 1 at 0x404828: file pg_repack.c, line 1101.

gdb) r
Starting program: /home/dmitry/Projects/pg_repack/bin/pg_repack -t test db
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

# ok, struct 'table' should be filled with Oids etc
Breakpoint 1, repack_one_table (table=0x7fffffffdc80, orderby=0x0) at pg_repack.c:1101
1101            PGresult           *res = NULL
```

Step 2:
```plpgsql
-- terminal B

 -- no lock on table 'test'
select locktype, relation::regclass, mode from pg_locks ;
  locktype  | relation |      mode       
------------+----------+-----------------
 relation   | pg_locks | AccessShareLock
 virtualxid |          | ExclusiveLock
(2 rows)

 -- we can drop it
drop table test;
DROP TABLE
```

Step 3:
```bash
# terminal A
(gdb) c
Continuing.
INFO: repacking table "test"
ERROR:  relation "test" does not exist
ERROR: query failed: ERROR:  current transaction is aborted, commands ignored until end of transaction block
DETAIL: query was: RESET statement_timeout
ERROR: query failed: ERROR:  table name not found for OID 1339794
DETAIL: query was: SELECT repack.repack_drop($1, $2)
ERROR: query failed: ERROR:  table name not found for OID 1339794
DETAIL: query was: SELECT repack.repack_drop($1, $2)
... (loop)
```

I'm sure that this problem should be addressed. Concurrent `DROP TABLE` is possible, so I decided to check the number of result tuples at least in code I've changed in my patch.

We should add a few checks here and there (probably in `repack_one_table()` too), or take locks beforehand. I didn't dare to make such changes.